### PR TITLE
UGENE-7584 Spidey alignment fails if delete created annotation until …

### PR DESCRIPTION
**Pull request by DmitriiSukhomlinov**
…process is finished

### CODE REVIEW

**Review by yalgaer — CHANGES_REQUESTED** [review]
We should avoid showing dialogs/message boxes from within TaskScheduler callbacks: such dialogs stop the event loop - it means that no other tasks are processed while the dialog is shown at all - all tasks are blocked.

Even worse - this may lead to a dead-locks: when a task is waited to be finished inside of the event loop created by a the dialog started in prepare/report/onSubtaskFinished etc..

An acceptable simple solution might be either canceling the task completely or creating a new annotation table with an auto-generated name.

More complex option is to handle the task results outside of the TaskScheduler callbacks: wait for the task to finish and run results processing code (with dialogs) later.

**Comment by DmitriiSukhomlinov** [comment]
> We should avoid showing dialogs/message boxes from within TaskScheduler callbacks: such dialogs stop the event loop - it means that no other tasks are processed while the dialog is shown at all - all tasks are blocked.
> 
> Even worse - this may lead to a dead-locks: when a task is waited to be finished inside of the event loop created by a the dialog started in prepare/report/onSubtaskFinished etc..
> 
> An acceptable simple solution might be either canceling the task completely or creating a new annotation table with an auto-generated name.
> 
> More complex option is to handle the task results outside of the TaskScheduler callbacks: wait for the task to finish and run results processing code (with dialogs) later.

I agree, fixed

File: src/plugins_3rdparty/primer3/src/primer3_core/primer3.h [hunk 1]
```diff
@@ -176,7 +176,7 @@ typedef struct primrec {
  * returns a pair of primers.)
  */
 typedef struct primpair {
-  double pair_quality;
```
Comment (yalgaer) [inline src/plugins_3rdparty/primer3/src/primer3_core/primer3.h]: Please remove unrelated changes.

File: src/plugins/external_tool_support/src/spidey/SpideySupportTask.cpp [hunk 2]
```diff
@@ -221,24 +225,46 @@ void PrepareInputForSpideyTask::run() {
...
-        return res;
+        return {};
     }
 
     if (subTask == spideyAlignmentTask) {
+        if (aObj == nullptr) {
+            taskLog.details("The result of the \"SpideySupportTask\" should be saved to the annotation table, which has been removed. The new annotation table is about to be created.");
+
+            auto dnaSeq = settings.getCDnaSequence();
+            auto documentFilePath = GUrlUtils::getDefaultDataPath() + "/MyDocument.gb";
```
Comment (yalgaer) [inline src/plugins/external_tool_support/src/spidey/SpideySupportTask.cpp:249]: Please roll the result file name or we can overwrite user data.

**Review by yalgaer — CHANGES_REQUESTED** [review]
Minor issues only. Overall the solution is correct.

File: src/plugins/external_tool_support/src/spidey/SpideySupportTask.cpp (hunk 2 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/external_tool_support/src/spidey/SpideySupportTask.cpp:249]: You won't override the user data in any case - even if you write to the existed "MyDocument.gb" and it has an annotation with the same name. There is a point, that you create the same document as a default one.

File: src/plugins_3rdparty/primer3/src/primer3_core/primer3.h (hunk 1 above)
Comment (DmitriiSukhomlinov) [reply src/plugins_3rdparty/primer3/src/primer3_core/primer3.h]: Fixed

File: src/plugins/external_tool_support/src/spidey/SpideySupportTask.cpp (hunk 2 above)
Comment (yalgaer) [reply src/plugins/external_tool_support/src/spidey/SpideySupportTask.cpp:249]: This is how I see the problem: 

1. You have an existing document that is not added to the project: /var/MyDocument.gb
2. You create the same in-memory document with a new data and add it to the project as 'loaded': /var/MyDocument.gb

What will happen next? 
I bet the newly created document will override the old one and this means a data loss for a user?

**Review by yalgaer — CHANGES_REQUESTED** [review]
Please check the comment about the document overwrite

File: src/plugins/external_tool_support/src/spidey/SpideySupportTask.cpp (hunk 2 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/external_tool_support/src/spidey/SpideySupportTask.cpp:249]: Fixed

**Review by yalgaer — COMMENTED** [review]
Good, will approve after GUI tests.

**Comment by DmitriiSukhomlinov** [comment]
@yalgaer all tests have passed

**Review by yalgaer — APPROVED** [review]

**Pull request merged**